x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/build_exceptions_query.ts

```diff
@@ -0,0 +1,151 @@
+
+import { Query } from '../../../../../../../../src/plugins/data/server';
+import { List, ListOperator, ListValues } from '../routes/schemas/types/lists_default_array';
+import { RuleAlertParams } from '../types';
+
+export const getLanguageBooleanOperator = (language: string, value: string) => {
+  if (language === 'lucene') {
+    return value.toUpperCase();
+  }
+
+  return value;
+};
+
+export const operatorBuilder = (operator: ListOperator, language: string): string => {
+  return operator === 'excluded'
+    ? ` ${getLanguageBooleanOperator(language, 'and')} `
+    : ` ${getLanguageBooleanOperator(language, 'and')} ${getLanguageBooleanOperator(
+        language,
+        'not'
+      )} `;
+};
+
+export const buildExists = ({
+  operator,
+  field,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  language: string;
+}): string => {
+  switch (language) {
+    case 'kuery':
+      return `${operatorBuilder(operator, language)}${field}:*`;
+    case 'lucene':
+      return `${operatorBuilder(operator, language)}_exists_${field}`;
+    default:
+      return '';
+  }
+};
+
+export const buildMatch = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  const value = values[0].name;
+  return `${operatorBuilder(operator, language)}${field}:${value}`;
+};
+
+export const buildMatchAll = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  if (values.length === 1) {
+    return buildMatch({ operator, field, values, language });
+  } else {
+    const matchAllValues = values.map(value => {
+      return value.name;
+    });
+
+    return `${operatorBuilder(operator, language)}${field}:(${matchAllValues.join(
+      ` ${getLanguageBooleanOperator(language, 'or')} `
+    )})`;
+  }
+};
+
+export const evaluateValues = ({ list, language }: { list: List; language: string }): string => {
+  const { values_operator: operator, values_type: type, field, values } = list;
+  switch (type) {
+    case 'exists':
+      return buildExists({ operator, field, language });
+    case 'match':
+      return buildMatch({ operator, field, values: values!, language });
+    case 'match_all':
+      return buildMatchAll({ operator, field, values: values!, language });
+    default:
+      return '';
+  }
+};
+
+export const buildExceptions = ({
+  query,
+  lists,
+  language,
+  accumulator = [],
+}: {
+  query: string;
+  lists: List[];
+  language: string;
+  accumulator?: string[];
+}): string[] => {
+  return lists.reduce((acc, item) => {
+    const exception: string[] = [query];
+    const { and, ...exceptionDetails } = { ...item };
+
+    exception.push(evaluateValues({ list: exceptionDetails, language }));
+
+    if (and && !isEmpty(and)) {
+      and.forEach(i => {
+        exception.push(evaluateValues({ list: i, language }));
```

Comment:
Here you can probably use a `map` in conjunction with a spread to add a new value to the exception without resorting to the `forEach`. Using `map` to get a new value out and a new reference and then appending that to the end of the exception (if possible) is generally preferable over using mutatious push.

Just guessing, something like this:
```ts
    if (and != null) {
      const andException = and.map(i => evaluateValues({ list: i, language }));
      exception = [...exception, ...andException];
    }
```

As an aside I don't usually worry if the array or a new shallow copy because of an empty array occurs as it's cheap and the more new references to stop people from relying on any side effect references the better is why I usually allow empty arrays to pass through.

Reply:
Yea, I didn't feel great about using `forEach` here - thanks for calling that out. I ended up just calling `buildExceptions` and added a param to denote when to include the query in the final output in addition to using the spread operator as you suggested.